File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where

import Data.Char(isUpper, isSpace, isLower)

responseFor :: String -> String
responseFor xs
  | isSilent xs = "Fine. Be that way!"
  | allUpper xs && isQuestion xs = "Calm down, I know what I'm doing!"
  | isQuestion xs = "Sure."
  | allUpper xs = "Whoa, chill out!"
  | otherwise = "Whatever."
  where
    allUpper xs = any isUpper xs && not (any isLower xs)
    isQuestion xs = not (null xs) && last (filter (not. isSpace) xs) == '?'
    isSilent xs = null xs || all isSpace xs

File: solutions/julia/bob/1/bob.jl
function bob(stimulus)
    cleaned = strip(stimulus)
    if isempty(cleaned)
        return "Fine. Be that way!"
    elseif cleaned[end] == '?'
        if uppercase(cleaned) == cleaned && uppercase(cleaned) != lowercase(cleaned) return "Calm down, I know what I'm doing!" end
        return "Sure."
    elseif uppercase(cleaned) == cleaned && uppercase(cleaned) != lowercase(cleaned)
        return "Whoa, chill out!"
    end
    return "Whatever."
end
